converted the slave to master access control to github (#3317)
jenkins-infra/jenkins.io: content/doc/book/using/agent-master-access-control.adoc

those commands, and wait for project developers to perform this vetting
process. Once we verified that they are safe, you can whitelist them by
using this mechanism.

[[AgentToMasterAccessControl-Fileaccessrules]]
=== File access rules

File access request from agents is tested against the rules you specify.
Each rule is a tuple that consists of:

* *allow/deny*: if the following two parameters match the current
request being considered, an "allow" entry would allow the request to be
carried out and a "deny" entry would deny the request to be rejected,
regardless of what later rules might say.
* *operation*: the type of the operation requested. The following 6
values exist. You can also list them separating with ',' or use "all" to
indicate a match for all operations:
** read: read file content or list directory entries
** write: write file content
** mkdirs: create a new directory
** create: create a file in an existing directory
** delete: delete a file or directory
** stat: read metadata of a file/directory, such as timestamp, length,
file access modes.
* *file path*: regular expression that specifies file paths that match
this rule. In addition to

> http://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html[the

Reviewer: Let's use the Java 8 javadoc since that is the minimum Java version now supported by Jenkins.
```suggestion
http://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html[the
```

[source,groovy]
----
// UNSAFE
class SomeCodeThatRunsOnAgent {
    void readBackSomeFileFromMaster() {
        final String path = "...";
        channel.call(new SlaveToMasterCallable<String,IOException>() {
            public String call() {
                return FileUtils.readFileToString(new File(path));
            }
        });
    }
}
----

`+Callable+` that delegates execution to a deserialized object is
dangerous and needs to be carefully examined, because a malicious agent
can designate unintended `+Runnable+` object:

[source,groovy]
----
// UNSAFE
class MyCallable extends SlaveToMasterCallable<Void> {
    Runnable r;
    public Void call() {
        r.run();
        return null;
    }
}
----

To avoid this hassle entirely, consider rewriting your code not to call
back to a master from an agent. Instead, when a master first sends a
command to an agent, you can carry all the data you'll need with you.
This may not be always possible or practical, but it's a lot easier to
secure.

[[AgentToMasterAccessControl-FixingpluginswithoutrequiringnewerJenkins]]
=== Fixing plugins without requiring newer Jenkins

> Classifying `+Callable+`/`+FileCallable+` requires new classes added to
> Jenkins 1.587/1.580.1. This poses a challenge if you want to retain
> backward compatibility with earlier versions of Jenkins.
> 

Reviewer: ```suggestion
Classifying `+Callable+`/`+FileCallable+` requires new classes added to Jenkins 1.580.1. 
Plugins should require a Jenkins base version that is significantly newer than that minimum requirement.

```

> To solve this problem, we've developed
> https://github.com/jenkinsci/SECURITY-144-compat[SECURITY-144-compat]
> module. This module let you classify `+Callable+`, while still
> functioning correctly on earlier versions of Jenkins. See the
> https://github.com/jenkinsci/SECURITY-144-compat/blob/master/README.md[documentation
> of `+SECURITY-144-compat+`] for details.
> 
> Note
> 
> As of version 1.1, this library is deprecated, as its use caused some
> unresolved problems
> (https://issues.jenkins-ci.org/browse/JENKINS-25625[JENKINS-25625]). +
> Anyway 1.580.1 is now a fairly conservative choice of baseline: you will
> not exclude so many users by requiring it for new plugin releases.

Reviewer: Remove paragraphs that describe ancient history and a workaround for ancient history.
```suggestion
```
